File: src/main/kotlin/io/github/tmarsteel/flyingnarrator/nefs/NefsCommand.kt
@file:OptIn(ExperimentalSerializationApi::class)
@file:UseSerializers(SystemPathSerializer::class)
package io.github.tmarsteel.flyingnarrator.nefs

import io.github.tmarsteel.flyingnarrator.io.SystemPathSerializer
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Polymorphic
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.json.JsonClassDiscriminator
import java.nio.file.Path

@Serializable
@Polymorphic
@JsonClassDiscriminator("op")
sealed interface NefsCommand {
    @Serializable
    @SerialName("ls")
    class EnumerateFiles : NefsCommand {
        @Serializable
        data class Result(
            val files: List<String>,
        )
    }

    @Serializable
    @SerialName("extract")
    class Extract(
        val pathInNefs: String,
        val extractTo: Path,
        val decodeBinaryXml: Boolean,
    ) : NefsCommand {

        @Serializable
        data class Result(
            val itemLocated: Boolean,
            val extractionSuccessful: Boolean,
        )
    }

    @Serializable
    @SerialName("exit")
    class Exit : NefsCommand
}

File: nefsedit-cli/nefsedit-cli/Interface/NefsCommand.cs
using System.Text.Json.Serialization;

namespace NefsEditCLI.Interface;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "op")]
[JsonDerivedType(typeof(Enumerate), typeDiscriminator: "ls")]
[JsonDerivedType(typeof(Extract), typeDiscriminator: "extract")]
[JsonDerivedType(typeof(Exit), typeDiscriminator: "exit")]
public abstract class NefsCommand
{
    public class Enumerate : NefsCommand
    {
        public class Result(List<string> files)
        {
            public List<string> files { get; set; } = files;
        }
    }

    public class Extract : NefsCommand
    {
        public string pathInNefs { get; set; }
        public string extractTo { get; set; }
        public bool decodeBinaryXml { get; set; } = false;
        
        public class Result(bool itemLocated, bool extractionSuccessful)
        {
            public bool itemLocated { get; set; } = itemLocated;
            public bool extractionSuccessful { get; set; } = extractionSuccessful;
        }
    }

    public class Exit : NefsCommand
    {
        
    }
}